src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs

```diff
@@ -37,4 +40,43 @@ public JsonObject GetOrAdd(OpenApiSchemaKey key, Func<OpenApiSchemaKey, JsonObje
     {
         return _schemas.GetOrAdd(key, valueFactory);
     }
+
+    /// <summary>
+    /// Add the provided schema to the schema-with-references cache that is eventually
+    /// used to populate the top-level components.schemas object. This method will
+    /// unwrap the provided schema and add any child schemas to the global cache. Child
+    /// schemas include those referenced in the schema.Items, schema.AdditionalProperties, or
+    /// schema.Properties collections. A unique identifier is generated for each schema
+    /// on the update step to avoid populating the cache with schemas that only appear once in
+    /// the document.
+    /// </summary>
+    /// <param name="schema">The <see cref="OpenApiSchema"/> to add to the schemas-with-references cache.</param>
+    public void PopulateSchemaIntoReferenceCache(OpenApiSchema schema)
+    {
+        _schemasWithReference.AddOrUpdate(schema, (_) => null, (_, _) => Guid.NewGuid().ToString());
+        if (schema.AdditionalProperties is not null)
+        {
+            _schemasWithReference.AddOrUpdate(schema.AdditionalProperties, (_) => null, (_, _) => Guid.NewGuid().ToString());
```

Comment:
I saw the note that the guids are equal. How will the name be calculated if two types result in the same schema?

For example,
```cs
public class Account
{
    public int Id { get; set; }
    public string Name { get; set; }
}

public class Product
{
    public int Id { get; set; }
    public string Name { get; set; }
}
```
If I was looking at the OpenAPI JSON or in SwaggerUI, I would expect those to be two different schemas, e.g. `components/schemas/Account` and `components/schemas/Product`

Reply:
Yes, this is an unfortunate junction where the expectations of JSON Schema and the usability of OpenAPI don't mesh really well. :/

To balance this out, I'm materializing an `x-schema-id` property into the generated schemas with the friendly type name embedded (e.g. `Account` vs `Product`) so that they don't pass the equality checks here despite having the same properties defined.